contracts/Gateway/NativeTokenGateway.sol

```diff
@@ -0,0 +1,232 @@
+     * @custom:error ZeroAddressNotAllowed is thrown if address of minter is zero address
+     * @custom:error ZeroValueNotAllowed is thrown if mintAmount is zero
+     * @custom:event TokensWrappedAndSupplied is emitted when assets are supplied to the market
+     */
+    function wrapAndSupply(address minter) external payable nonReentrant {
+        ensureNonzeroAddress(minter);
+
+        uint256 mintAmount = msg.value;
+        ensureNonzeroValue(mintAmount);
+
+        wNativeToken.deposit{ value: mintAmount }();
+        IERC20(address(wNativeToken)).forceApprove(address(vWNativeToken), mintAmount);
+
+        vWNativeToken.mintBehalf(minter, mintAmount);
+
+        IERC20(address(wNativeToken)).forceApprove(address(vWNativeToken), 0);
+        emit TokensWrappedAndSupplied(minter, address(vWNativeToken), mintAmount);
+    }
+
+    /**
+     * @notice Redeem vWNativeToken, unwrap to Native Token, and send to the user
+     * @param redeemAmount The amount of underlying tokens to redeem
+     * @custom:error ZeroValueNotAllowed is thrown if redeemAmount is zero
+     * @custom:event TokensRedeemedAndUnwrapped is emitted when assets are redeemed from a market and unwrapped
+     */
+    function redeemUnderlyingAndUnwrap(uint256 redeemAmount) external nonReentrant {
+        _redeemAndUnwrap(redeemAmount, true);
+    }
+
+    /**
+     * @notice Redeem vWNativeToken, unwrap to Native Token, and send to the user
+     * @param redeemTokens The amount of vWNative tokens to redeem
+     * @custom:error ZeroValueNotAllowed is thrown if redeemTokens is zero
+     * @custom:event TokensRedeemedAndUnwrapped is emitted when assets are redeemed from a market and unwrapped
+     */
+    function redeemAndUnwrap(uint256 redeemTokens) external nonReentrant {
+        _redeemAndUnwrap(redeemTokens, false);
+    }
+
+    /**
+     * @dev Borrow wNativeToken, unwrap to Native, and send to the user
+     * @param borrowAmount The amount of underlying tokens to borrow
+     * @custom:error ZeroValueNotAllowed is thrown if borrowAmount is zero
+     * @custom:event TokensBorrowedAndUnwrapped is emitted when assets are borrowed from a market and unwrapped
+     */
+    function borrowAndUnwrap(uint256 borrowAmount) external nonReentrant {
+        ensureNonzeroValue(borrowAmount);
+
+        vWNativeToken.borrowBehalf(msg.sender, borrowAmount);
+
+        wNativeToken.withdraw(borrowAmount);
+        _safeTransferNativeTokens(msg.sender, borrowAmount);
+        emit TokensBorrowedAndUnwrapped(msg.sender, address(vWNativeToken), borrowAmount);
+    }
+
+    /**
+     * @notice Wrap Native, repay borrow in the market, and send remaining Native to the user
+     * @custom:error ZeroValueNotAllowed is thrown if repayAmount is zero
+     * @custom:event TokensWrappedAndRepaid is emitted when assets are repaid to a market and unwrapped
+     */
+    function wrapAndRepay() external payable nonReentrant {
```

Comment:
My knowledge in Solidity is very limited, but am I correct assuming this function does not allow repaying a full loan (since accrued interests need to be token in account)?

From a frontend perspective, the way we currently repay entire loans for other markets is by passing `MAX_UINT256` to the `repayBorrow` of the corresponding VToken contract. Perhaps we could add the same logic here?